src/pwr.rs

```diff
@@ -1,13 +1,145 @@
+    WKUP1,
+    #[doc = "Wake-up button 2"]
+    WKUP2,
+    #[doc = "Wake-up button 3"]
+    WKUP3,
+    #[doc = "Wake-up button 4"]
+    WKUP4,
+    #[doc = "Wake-up button 5"]
+    WKUP5,
+}
 
 pub struct Pwr {
     pub cr1: CR1,
     pub cr2: CR2,
     pub cr3: CR3,
     pub cr4: CR4,
+    pub scr: SCR,
+    pub sr1: SR1,
+}
+
+impl Pwr {
+    /// Configures dynamic voltage regulator range
+    ///
+    /// Will panic if low-power range is selected for higher system clock
+    pub fn set_power_range(&mut self, range: VosRange, clocks: &Clocks) -> Result<(), Error> {
+        match range {
+            VosRange::HighPerformance => unsafe {
+                {
+                    self.cr1
+                        .reg()
+                        .modify(|_, w| w.vos().bits(VosRange::HighPerformance as u8))
+                }
+                Ok(())
+            },
+            VosRange::LowPower => {
+                if clocks.sysclk() > 26.mhz().into() {
+                    Err(Error::SysClkTooHighVos)
+                } else {
+                    unsafe {
+                        self.cr1
+                            .reg()
+                            .modify(|_, w| w.vos().bits(VosRange::LowPower as u8))
+                    }
+                    Ok(())
+                }
+            }
+        }
+    }
+
+    /// Switches the system into low power run mode
+    pub fn low_power_run(&mut self, clocks: &Clocks) -> Result<(), Error> {
+        if clocks.sysclk() > 2.mhz().into() {
+            Err(Error::SysClkTooHighLpr)
+        } else {
+            self.cr1.reg().modify(|_, w| w.lpr().set_bit());
+            Ok(())
+        }
+    }
+
+    /// Enters 'Shutdown' low power mode.
+    pub fn shutdown(&mut self, wkup: &[WakeUpSource], scb: &mut SCB) -> ! {
+        for source in wkup {
+            match source {
+                WakeUpSource::Internal => self.cr3.reg().modify(|_, w| w.ewf().set_bit()),
+                WakeUpSource::WKUP1 => self.cr3.reg().modify(|_, w| w.ewup1().set_bit()),
+                WakeUpSource::WKUP2 => self.cr3.reg().modify(|_, w| w.ewup2().set_bit()),
+                WakeUpSource::WKUP3 => self.cr3.reg().modify(|_, w| w.ewup3().set_bit()),
+                WakeUpSource::WKUP4 => self.cr3.reg().modify(|_, w| w.ewup4().set_bit()),
+                WakeUpSource::WKUP5 => self.cr3.reg().modify(|_, w| w.ewup5().set_bit()),
+            }
+        }
+        scb.set_sleepdeep();
+        self.scr.reg().write(|w| {
+            w.wuf1()
+                .set_bit()
+                .wuf2()
+                .set_bit()
+                .wuf3()
+                .set_bit()
+                .wuf4()
+                .set_bit()
+                .wuf5()
+                .set_bit()
+                .sbf()
+                .set_bit()
+        });
+        unsafe { self.cr1.reg().modify(|_, w| w.lpms().bits(0b111)) };
+        cortex_m::asm::dsb();
+        cortex_m::asm::wfi();
+        loop {}
```

Comment:
Why is the loop a part of this method? Would this not be up to the user to handle?

Reply:
This particular sleep mode is non-returnable, MCU will reset on wake-up and that function will never return from the `wfi()` call. That's why this  fuction is explicitly marked as diverging and loop just ensures compiler, that it is a diverging function.